spring-cloud-dataflow-core/src/main/java/org/springframework/cloud/dataflow/core/dsl/AbstractTokenizer.java

```diff
@@ -0,0 +1,330 @@
+/*
+ * Copyright 2015-2017 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.springframework.cloud.dataflow.core.dsl;
+
+import java.util.ArrayList;
+import java.util.List;
+
+import org.springframework.util.Assert;
+
+/**
+ * The base class for tokenizers. Knows about tokenization in general but
+ * not specific to a DSL variant. Provides state management and utility methods
+ * to real tokenizer subclasses.
+ *
+ * @author Andy Clement
+ */
+public abstract class AbstractTokenizer {
+
+	private final int[] NO_LINEBREAKS = new int[0];
+	
+	/**
+	 * The string to be tokenized.
+	 */
+	protected String expressionString;
+
+	/**
+	 * The expressionString as a char array.
+	 */
+	protected char[] toProcess;
+
+	/**
+	 * Length of input data.
+	 */
+	protected int max;
+
+	/**
+	 * Current lexing position in the input data.
+	 */
+	protected int pos;
+
+	/**
+	 * Output stream of tokens.
+	 */
+	protected List<Token> tokens = new ArrayList<Token>();
+	
+	/**
+	 * Positions of linebreaks in the parsed string.
+	 */
+	protected int[] linebreaks = NO_LINEBREAKS;
+
+	abstract void process();
+	
+	public Tokens getTokens(String inputData) {
+		this.expressionString = inputData;
+		this.toProcess = (inputData + "\0").toCharArray();
+		this.max = toProcess.length;
+		this.pos = 0;
+		this.tokens.clear();
+		process();
+		return new Tokens(inputData, tokens, linebreaks);
```

Comment:
I don't think we enforce this style (which I don't personally like, full disclosure) throughout our codebase(s)